Implement Priority Hints

rwlbuis opened the pull request:
#### 69a182bca140bcc4d604dd6eba4ea48a16e9d00a
<pre>
Implement Priority Hints
<a href="https://bugs.webkit.org/show_bug.cgi?id=252739">https://bugs.webkit.org/show_bug.cgi?id=252739</a>

Reviewed by Youenn Fablet.

Implement Priority Hints [1], specifically covering:
- processing the priority hint passed using fetch API [2]
- processing the priority hint for images
- processing the priority hint for scripts, both classic and module
- processing the priority hint for the link element
- processing the priority hint for the link HTTP header
- processing the priority hint in the html resource preloader

Tests are added covering this, including both dynamically inserting
elements (script/image/link) as well as static html.

The actual priority hint functionality consists of decreasing/increasing the
computed priority in the request for fetchpriority=low/high respectively, clamping
to the very low/very high limits.

[1] <a href="https://html.spec.whatwg.org/#fetch-and-process-the-linked-resource">https://html.spec.whatwg.org/#fetch-and-process-the-linked-resource</a>
[2] <a href="https://fetch.spec.whatwg.org/#dom-requestinit-priority">https://fetch.spec.whatwg.org/#dom-requestinit-priority</a>

* LayoutTests/http/tests/priority-hints/fetch-api-priority-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/fetch-api-priority.html: Added.
* LayoutTests/http/tests/priority-hints/image-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/image-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/image-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/image-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/link-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/link-header-fetch-priority-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-header-fetch-priority.py: Added.
* LayoutTests/http/tests/priority-hints/link-preload-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-preload-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/link-preload-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-preload-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/link-stylesheet-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-stylesheet-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-async-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-async-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/script-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-module-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-module-initial-load.html: Added.
* LayoutTests/http/tests/resources/checkResourcePriority.js: Added.
(checkResourcePriority):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchLoader.cpp:
(WebCore::FetchLoader::start):
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::buildOptions):
(WebCore::FetchRequest::initializeOptions):
(WebCore::FetchRequest::initializeWith):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/Modules/fetch/RequestPriority.h:
* Source/WebCore/bindings/js/CachedScriptFetcher.cpp:
(WebCore::CachedScriptFetcher::requestScriptWithCache const): pass priority hint to the cached resource request
* Source/WebCore/bindings/js/CachedScriptFetcher.h:
(WebCore::CachedScriptFetcher::CachedScriptFetcher):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::InlineClassicScript):
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::LoadableNonModuleScriptBase):
(WebCore::LoadableClassicScript::create):
(WebCore::LoadableClassicScript::LoadableClassicScript):
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableImportMap.cpp:
(WebCore::LoadableImportMap::LoadableImportMap):
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::create):
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.h:
(WebCore::LoadableScript::LoadableScript):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::fetchPriorityHint const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::ScriptElementCachedScriptFetcher):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::fetchPriorityForBindings const):
(WebCore::HTMLImageElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::process): pass priority hint to the cached resource request
(WebCore::HTMLLinkElement::fetchPriorityForBindings const):
(WebCore::HTMLLinkElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::fetchPriorityForBindings const):
(WebCore::HTMLScriptElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::createPreloadRequest):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
* Source/WebCore/html/parser/HTMLResourcePreloader.cpp: handle fetch priority in preloading for image/script/link
(WebCore::PreloadRequest::resourceRequest): pass priority hint to the cached resource request
* Source/WebCore/html/parser/HTMLResourcePreloader.h:
(WebCore::PreloadRequest::PreloadRequest):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement): pass priority hint to the cached resource request
* Source/WebCore/loader/LinkHeader.cpp: parse fetchpriority link header
(WebCore::parameterNameFromString):
(WebCore::LinkHeader::setValue):
* Source/WebCore/loader/LinkHeader.h:
(WebCore::LinkHeader::fetchPriorityHint const):
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::loadLinksFromHeader):
(WebCore::LinkLoader::preloadIfNeeded): pass priority hint to the cached resource request
* Source/WebCore/loader/LinkLoader.h:
* Source/WebCore/loader/ResourceLoaderOptions.h:
(WebCore::ResourceLoaderOptions::fetchPriorityHint):
(WebCore::ResourceLoaderOptions::ResourceLoaderOptions):
(WebCore::ResourceLoaderOptions::loadedFromPluginElement): Deleted.
* Source/WebCore/loader/ThreadableLoader.cpp:
(WebCore::ThreadableLoaderOptions::isolatedCopy const):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource):
(WebCore::CachedResource::setLoadPriority): actually calculate final priority taking the hint into account
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::fetchPriorityHint const):
* Source/WebCore/testing/Internals.cpp: add API to get the priority used in the resource request
(WebCore::toInternalsResourceLoadPriority):
(WebCore::Internals::getResourcePriority):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/261689@main">https://commits.webkit.org/261689@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d163f99f35f7b9c1e0eda0011b8be6f3e220475

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21708 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1213 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4332 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121111 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23046 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12869 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5475 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118324 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100321 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105624 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99066 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46129 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14045 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/903 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14727 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20067 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52908 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8147 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16564 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/html/HTMLScriptElement.cpp:
```diff
@@ -212,7 +212,12 @@ void HTMLScriptElement::setFetchPriorityForBindings(const AtomString& value)
+    return convertEnumerationToString(fetchPriorityHint());
+}
+
+RequestPriority HTMLScriptElement::fetchPriorityHint() const
```
Could we move this one to HTMLElement, maybe fetchPriorityForBindings as well?

rwlbuis replied:
That makes sense, though I think referrerPolicy has the same organisation. Does it makes sense to fix both in a follow up patch?

rwlbuis replied:
I looked into this a bit more. fetchPriorityHint is also used as virtual in ScriptElement so I think this is not worth moving for two elements.

rwlbuis replied:
Regarding fetchPriorityForBindings, that does seem to make sense, but we have the same structure for referrerPolicy and loading, so I think it is better to do that in a follow up patch.

youennf commented on Source/WebCore/loader/FetchOptions.h:
```diff
@@ -62,6 +63,7 @@ struct FetchOptions {
     Cache cache { Cache::Default };
     Redirect redirect { Redirect::Follow };
     ReferrerPolicy referrerPolicy { ReferrerPolicy::EmptyString };
+    RequestPriority fetchPriorityHint { RequestPriority::Auto };
```
Given it is not exposed in service worker, I would move it to ResourceLoaderOptions.

rwlbuis replied:
Done.

youennf commented on Source/WebCore/loader/cache/CachedResourceRequest.h:
```diff
@@ -121,6 +124,7 @@ class CachedResourceRequest {
     String m_charset;
     ResourceLoaderOptions m_options;
     std::optional<ResourceLoadPriority> m_priority;
+    RequestPriority m_fetchPriorityHint;
```
Not needed since it is stored in m_options.

rwlbuis replied:
Removed.

youennf commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -991,6 +991,32 @@ void Internals::setStrictRawResourceValidationPolicyDisabled(bool disabled)
+
+std::optional<Internals::ResourceLoadPriority> Internals::getResourcePriority(const String& url)
+{
+    Document* document = contextDocument();
```
auto* and return early if there is no document

rwlbuis replied:
Done.

youennf replied:
It does not appear to be done

rwlbuis replied:
Sorry, I was confused which line should gain the auto.

youennf commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -896,12 +896,21 @@ unsigned CachedResource::overheadSize() const
+        priority = loadPriority.value();
     else
-        m_loadPriority = DefaultResourceLoadPriority::forResourceType(type());
+        priority = DefaultResourceLoadPriority::forResourceType(type());
```
one liner:
m_loadPriority = loadPriority ? * loadPriority : ...
I am not sure priority var is need.

rwlbuis replied:
Done the conditional, however I think the temp var is needed since m_loadPriority is a bitfield.

youennf commented on Source/WebCore/loader/cache/CachedResourceRequest.h:
```diff
@@ -64,6 +64,9 @@ class CachedResourceRequest {
     void setPriority(std::optional<ResourceLoadPriority>&& priority) { m_priority = WTFMove(priority); }
 
+    RequestPriority fetchPriorityHint() const { return m_fetchPriorityHint; }
+    void setFetchPriorityHint(const RequestPriority& fetchPriorityHint) { m_fetchPriorityHint = fetchPriorityHint; }
```
s/const RequestPriority&/RequestPriority/
s/m_fetchPriorityHint/m_options. fetchPriorityHint/

rwlbuis replied:
Done.

youennf commented on Source/WebCore/loader/LinkLoader.h:
```diff
@@ -54,6 +54,7 @@ struct LinkLoadParameters {
     String imageSizes;
     String nonce;
     ReferrerPolicy referrerPolicy { ReferrerPolicy::EmptyString };
+    RequestPriority fetchPriorityHint;
```
Probably does not hurt to initialize it.

rwlbuis replied:
Sure, done.

youennf commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -212,6 +213,7 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
         resourceRequest.setInspectorInitiatorNodeIdentifier(InspectorInstrumentation::identifierForNode(m_element));
 
         auto request = createPotentialAccessControlRequest(WTFMove(resourceRequest), WTFMove(options), document, crossOriginAttribute);
+        request.setFetchPriorityHint(options.fetchPriorityHint);
```
Probably not needed since we are passing options when creating request.

rwlbuis replied:
I think so as well, removed.

youennf commented on Source/WebCore/Modules/fetch/FetchRequest.cpp:
```diff
@@ -92,6 +92,9 @@ static std::optional<Exception> buildOptions(FetchOptions& options, ResourceRequ
         options.referrerPolicy = init.referrerPolicy.value();
 
+    if (init.priority)
+        fetchPriorityHint = init.priority.value();
```
I would go with *init.priority()

rwlbuis replied:
I agree that looks nicer, fixed.

rwlbuis committed:
Implement Priority Hints
https://bugs.webkit.org/show_bug.cgi?id=252739

Reviewed by Youenn Fablet.

Implement Priority Hints [1], specifically covering:
- processing the priority hint passed using fetch API [2]
- processing the priority hint for images
- processing the priority hint for scripts, both classic and module
- processing the priority hint for the link element
- processing the priority hint for the link HTTP header
- processing the priority hint in the html resource preloader

Tests are added covering this, including both dynamically inserting
elements (script/image/link) as well as static html.

The actual priority hint functionality consists of decreasing/increasing the
computed priority in the request for fetchpriority=low/high respectively, clamping
to the very low/very high limits.

[1] https://html.spec.whatwg.org/#fetch-and-process-the-linked-resource
[2] https://fetch.spec.whatwg.org/#dom-requestinit-priority

* LayoutTests/http/tests/priority-hints/fetch-api-priority-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/fetch-api-priority.html: Added.
* LayoutTests/http/tests/priority-hints/image-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/image-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/image-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/image-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/link-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/link-header-fetch-priority-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-header-fetch-priority.py: Added.
* LayoutTests/http/tests/priority-hints/link-preload-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-preload-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/link-preload-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-preload-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/link-stylesheet-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/link-stylesheet-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-async-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-async-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-dynamic-insertion-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-dynamic-insertion.html: Added.
* LayoutTests/http/tests/priority-hints/script-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-initial-load.html: Added.
* LayoutTests/http/tests/priority-hints/script-module-initial-load-expected.txt: Added.
* LayoutTests/http/tests/priority-hints/script-module-initial-load.html: Added.
* LayoutTests/http/tests/resources/checkResourcePriority.js: Added.
(checkResourcePriority):
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchLoader.cpp:
(WebCore::FetchLoader::start):
* Source/WebCore/Modules/fetch/FetchRequest.cpp:
(WebCore::buildOptions):
(WebCore::FetchRequest::initializeOptions):
(WebCore::FetchRequest::initializeWith):
* Source/WebCore/Modules/fetch/FetchRequest.h:
* Source/WebCore/Modules/fetch/RequestPriority.h:
* Source/WebCore/bindings/js/CachedScriptFetcher.cpp:
(WebCore::CachedScriptFetcher::requestScriptWithCache const): pass priority hint to the cached resource request
* Source/WebCore/bindings/js/CachedScriptFetcher.h:
(WebCore::CachedScriptFetcher::CachedScriptFetcher):
* Source/WebCore/dom/InlineClassicScript.cpp:
(WebCore::InlineClassicScript::InlineClassicScript):
* Source/WebCore/dom/LoadableClassicScript.cpp:
(WebCore::LoadableNonModuleScriptBase::LoadableNonModuleScriptBase):
(WebCore::LoadableClassicScript::create):
(WebCore::LoadableClassicScript::LoadableClassicScript):
* Source/WebCore/dom/LoadableClassicScript.h:
* Source/WebCore/dom/LoadableImportMap.cpp:
(WebCore::LoadableImportMap::LoadableImportMap):
* Source/WebCore/dom/LoadableModuleScript.cpp:
(WebCore::LoadableModuleScript::create):
(WebCore::LoadableModuleScript::LoadableModuleScript):
* Source/WebCore/dom/LoadableModuleScript.h:
* Source/WebCore/dom/LoadableScript.h:
(WebCore::LoadableScript::LoadableScript):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::requestClassicScript):
(WebCore::ScriptElement::requestModuleScript):
* Source/WebCore/dom/ScriptElement.h:
(WebCore::ScriptElement::fetchPriorityHint const):
* Source/WebCore/dom/ScriptElementCachedScriptFetcher.h:
(WebCore::ScriptElementCachedScriptFetcher::ScriptElementCachedScriptFetcher):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::fetchPriorityForBindings const):
(WebCore::HTMLImageElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLLinkElement.cpp:
(WebCore::HTMLLinkElement::process): pass priority hint to the cached resource request
(WebCore::HTMLLinkElement::fetchPriorityForBindings const):
(WebCore::HTMLLinkElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLLinkElement.h:
* Source/WebCore/html/HTMLScriptElement.cpp:
(WebCore::HTMLScriptElement::fetchPriorityForBindings const):
(WebCore::HTMLScriptElement::fetchPriorityHint const):
* Source/WebCore/html/HTMLScriptElement.h:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::createPreloadRequest):
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
* Source/WebCore/html/parser/HTMLResourcePreloader.cpp: handle fetch priority in preloading for image/script/link
(WebCore::PreloadRequest::resourceRequest): pass priority hint to the cached resource request
* Source/WebCore/html/parser/HTMLResourcePreloader.h:
(WebCore::PreloadRequest::PreloadRequest):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::updateFromElement): pass priority hint to the cached resource request
* Source/WebCore/loader/LinkHeader.cpp: parse fetchpriority link header
(WebCore::parameterNameFromString):
(WebCore::LinkHeader::setValue):
* Source/WebCore/loader/LinkHeader.h:
(WebCore::LinkHeader::fetchPriorityHint const):
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::loadLinksFromHeader):
(WebCore::LinkLoader::preloadIfNeeded): pass priority hint to the cached resource request
* Source/WebCore/loader/LinkLoader.h:
* Source/WebCore/loader/ResourceLoaderOptions.h:
(WebCore::ResourceLoaderOptions::fetchPriorityHint):
(WebCore::ResourceLoaderOptions::ResourceLoaderOptions):
(WebCore::ResourceLoaderOptions::loadedFromPluginElement): Deleted.
* Source/WebCore/loader/ThreadableLoader.cpp:
(WebCore::ThreadableLoaderOptions::isolatedCopy const):
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::CachedResource):
(WebCore::CachedResource::setLoadPriority): actually calculate final priority taking the hint into account
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestResource):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::fetchPriorityHint const):
* Source/WebCore/testing/Internals.cpp: add API to get the priority used in the resource request
(WebCore::toInternalsResourceLoadPriority):
(WebCore::Internals::getResourcePriority):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/261689@main
